Remove unneeded paging when using PanacheQuery

robmv opened the pull request:
Avoid setting paging information on the native query implementation when no paging has been set on PanacheQuery.
 
Closes #8098

robmv commented:
The comment for the `page` field was removed because it doesn't apply any more, it is stored now because there are two states, by page and range, so both are stored.

I don't have an installation of MongoDB for testing but the same changes where done for it, and optimizing the query to limit 1 or 2 entries when using `firstResult*()` and `singleResult*()`

One extra question, currently you can switch between paging and range states, but it isn't possible to switch to non paging and non range states, unless you set null the page and the range, but that is an implementation leftover. Should an API for this be added, or document that page and range can be set to null, or add null checks to these APIs? (for another patch)

robmv committed:
Remove unneeded paging when using *PanacheQuery

Avoid setting paging information on the native query implementation when
no paging has been set on PanacheQuery and ReactivePanacheQuery

robmv commented:
Force updated because mixed accounts for the author of the commit

FroMage commented on extensions/panache/hibernate-orm-panache/runtime/src/main/java/io/quarkus/hibernate/orm/panache/runtime/PanacheQueryImpl.java:
```diff
@@ -100,11 +97,15 @@ public int pageCount() {
+        if (page == null) {
+            throw new UnsupportedOperationException("Cannot call a page related method, " +
+                    "call page(Page) or page(int, int) to initiate pagination first");
+        }
```
Well that's a change in behaviour: we used to have a single page encompassing all results. Perhaps it's fine, but at the very least, this has to be mentioned in the javadocs for the `PanacheQuery` interface.

robmv replied:
Javadoc updated, Noticed too that there is another interface `ReactivePanacheQuery` that has the same behaviour, so I updated it too so all three of them match.

FroMage reviewed (changes requested):
Thanks. We need some javadoc to indicate why the page methods can throw now that we don't have a default page.
We also need some tests to make sure we don't set paging info by default, though I'm not quite sure how to test that. Perhaps it's not possible.

robmv commented:
Just a of topic question, why these interfaces are defined this way?

```java
public interface PanacheQuery<Entity> {
    public <T extends Entity> PanacheQuery<T> page(Page page);
    ...
}
```

instead of:

```java
public interface PanacheQuery<Entity> {
    public PanacheQuery<Entity> page(Page page);
    ...
}
```

Why these methods have to change the returned generic from `PanacheQuery<Entity>` to `PanacheQuery<T extends Entity>`?. The only method that should do that is `project()`, all others do not modify the target entity. Am I missing something?

robmv commented:
> We also need some tests to make sure we don't set paging info by default, though I'm not quite sure how to test that. Perhaps it's not possible.

I agree that testing it will be difficult without going inside the wrapped APIs (JPA, MongoDB). The only way I think it is possible externally is to have a table with more than 2^31 rows, not very practical.

FroMage commented:
In RESTEasy they do some testing by examining the logs. Do you know if we can write a test that checks some logs @geoand @stuartwdouglas @dmlloyd ?

geoand commented:
The `QuarkusProdModeTest` does provide that ability:

For build logs see: https://github.com/quarkusio/quarkus/blob/master/integration-tests/spring-data-jpa/src/test/java/io/quarkus/it/spring/data/jpa/PropertyWarningsPMT.java#L32

For logs of the running application: https://github.com/quarkusio/quarkus/blob/master/integration-tests/kubernetes/standard/src/test/java/io/quarkus/it/kubernetes/KubernetesWithHealthTest.java#L51

FroMage commented:
Indeed in `QuarkusProdModeTest` you can get the `ProdModeTestResults.retainedBuildLogRecords` (you can inject `ProdModeTestResults`) and you can set `QuarkusProdModeTest.setLogRecordPredicate` to mark the hibernate SQL logs as being recorded so you can later obtain them.

FroMage commented:
Thanks @geoand !

geoand commented:
You are welcome @FroMage :)

robmv commented:
Thanks @geoand. @FroMage: I will check the suggestions and add test in the following days.

robmv commented:
Rebased in order to pull changes to ReactivePanacheQueryImpl on master.

I added a test based on the example `KubernetesWithHealthTest`. As the ProdModeTest run on another JVM I have to add the `prod-mode` surefire plugin, because there is already running and H2 server on normal tests and both conflicts.

I have to add  temporary workaround on `NoPagingPMT` to start the H2 server, because not matter what I do with the `QuarkusProdModeTest` instance, It doesn't start. I tried adding `H2DatabaseTestResource.class` with `addClasses()`  (Note that `TestResources.class` already has the `H2DatabaseTestResource` annotation and is added). Tried adding the artifact `quarkus-test-h2` with `setForcedDependencies()`. No luck. Any ideas?, maybe @geoand could help.

gsmet commented:
@robmv could you squash your commits into one?

@FroMage @geoand what's the status of this?

geoand commented:
It slipped through the cracks for me... Sorry.
I'll see if I can answer the question tomorrow

geoand commented:
@robmv I am not sure I understand the issue you are facing. Can you please elaborate?

FroMage commented on integration-tests/hibernate-orm-panache/src/test/java/io/quarkus/it/panache/NoPagingPMT.java:
```diff
@@ -0,0 +1,92 @@
+    static {
+        // WORKAROUND
+        try {
+            Server.createTcpServer().start();
```
This worries me a bit, why is it needed and where is it stopped?

robmv replied:
As ProdModeTest build a small JAR and run the test on a different JVM, There is a need to start the H2 Database server, My question to @geoand that I will detail later is about this, The workaround to start it manually is because the ProdModeTest isn't starting it even when I am adding `TestResources` that has the `@QuarkusTestResource(H2DatabaseTestResource.class)` annotation

As a temporary workaround didn't closed it myself because it is closed when that specific test JVM is closed.

geoand replied:
Yeah, the prod mode test doesn't handle resources...

geoand replied:
We could make it handle them for sure, but currently it doesn't.

FroMage commented on integration-tests/hibernate-orm-panache/pom.xml:
```diff
@@ -109,6 +119,30 @@
+                    The prod mode tests need to be part of a different execution to ensure that they don't mess with the standard tests.
+                    By adding this configuration we ensure that the maven surefire plugin will execute twice, one for the regular **/*Test.java
+                    tests (using the 'default-test' execution), and one for the prod mode tests (this 'prod-mode' execution)
+                    -->
```
Are we sure we need to split the test runs in two parts?

geoand replied:
Yes, this is needed because of classpath issues - I've encountered it before.

robmv replied:
and in this case it wasn't for the classpath but because the regular tests already start an H2 database instance, and the *PMT test run on its own JVM, so there are TCP port conflicts.

geoand replied:
Right, good find

FroMage reviewed (changes requested):
Sorry I took so long to review. I added some questions.

robmv commented:
> @robmv I am not sure I understand the issue you are facing. Can you please elaborate?

@geoand The test needs an H2 Database running. I am packing everything that is needed to run the test:

```java
.setArchiveProducer(() -> ShrinkWrap.create(JavaArchive.class)
        .addClasses(PageItem.class, TestResources.class, NoPagingTestEndpoint.class))
```

`TestResources` has the `@QuarkusTestResource(H2DatabaseTestResource.class)` annotation but it doesn't automatically start the database, so I added the workaround to start it manually before `QuarkusProdModeTest` is instantiated. I want to fix the need for the workaround to start the H2 database explicitly.

geoand commented:
Thanks for the explanation @robmv.

I'll try and get `@QuarkusTestResource` support for the ProdModeTest soon (hopefully this week).

robmv commented:
> Thanks for the explanation @robmv.
> 
> I'll try and get `@QuarkusTestResource` support for the ProdModeTest soon (hopefully this week).

Great @geoand, I will wait for it, meanwhile I will squash everything and rebase and push that when that lands. Thanks

geoand commented:
#8597 should take care of the `@QuarkusTestResource` use.

robmv commented:
Squashed and workaround removed

geoand commented:
Please rebase PR onto the latest master in order to pick up a CI fix

FroMage reviewed (approved):
LGTM, thanks!

FroMage commented:
@gsmet this will make it in 1.5, right, not 1.4?

gsmet commented:
Yeah, I would prefer it being part of a CR so more in 1.5.

I could be convinced though if someone had a strong opinion about it being needed for 1.4.

FroMage commented:
I think it's a cosmetic issue, only for Oracle, so it can probably wait.

gsmet commented:
In the end I decided to backport it as we have another PR that needs it.
